components/brave_wallet/browser/internal/hd_key_ed25519_slip23.cc

```diff
@@ -0,0 +1,250 @@
+};
+
+// https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-eddsa-05#section-5.1.5
+// requires scalar to follow this requirements 'The lowest 3 bits of the first
+// octet are cleared, the highest bit of the last octet is cleared, and the
+// second highest bit of the last octet is set'.
+// https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf expands
+// this requirement to `We admit only those k such that the third highest bit
+// of the last byte of k is zero`
+base::span<uint8_t, kSlip23ScalarSize> ClampScalarEd25519Bip32(
+    base::span<uint8_t, kSlip23ScalarSize> scalar) {
+  scalar[0] &= 0b1111'1000;  // The lowest 3 bits of the first
+                             // octet are cleared
+
+  scalar[31] &= 0b0111'1111;  // highest bit of the last octet is cleared and
+  scalar[31] &= 0b1101'1111;  // third highest bit of the last byte of k is zero
+
+  scalar[31] |= 0b0100'0000;  // the second highest bit of the last octet is set
+  return scalar;
+}
+
+// https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-eddsa-05#section-5.1.5
+// requires scalar to follow this requirements 'The lowest 3 bits of the first
+// octet are cleared, the highest bit of the last octet is cleared, and the
+// second highest bit of the last octet is set'.
+bool IsValidEd25519Scalar(base::span<const uint8_t, kSlip23ScalarSize> scalar) {
+  return (scalar[0] & 0b0000'0111) == 0b0000'0000 &&
+         (scalar[31] & 0b1100'0000) == 0b0100'0000;
+}
+
+std::array<uint8_t, 32> PubkeyFromScalar(base::span<const uint8_t, 32> scalar) {
+  DCHECK(IsValidEd25519Scalar(scalar));
+  std::array<uint8_t, 32> public_key;
+  ED25519_pubkey_from_scalar(public_key.data(), scalar.data());
+  return public_key;
+}
+
+DerivedHmacs DeriveHardenedHmacs(
+    uint32_t index,
+    base::span<const uint8_t, kSlip23ScalarSize> scalar,
+    base::span<const uint8_t, kSlip23PrefixSize> prefix,
+    base::span<const uint8_t> cc) {
+  std::array<uint8_t, 1 + 32 + 32 + 4> data;
+  auto span_writer = base::SpanWriter(base::span(data));
+  span_writer.Skip(1u);
+  span_writer.Write(scalar);
+  span_writer.Write(prefix);
+  span_writer.WriteU32LittleEndian(index);
+
+  DerivedHmacs result;
+
+  data[0] = 0x00;
+  result.z_hmac = crypto::hmac::SignSha512(cc, data);
+
+  data[0] = 0x01;
+  result.cc_hmac = crypto::hmac::SignSha512(cc, data);
+
+  return result;
+}
+
+DerivedHmacs DeriveNormalHmacs(
+    uint32_t index,
+    base::span<const uint8_t, kSlip23ScalarSize> scalar,
+    base::span<const uint8_t, kEd25519PublicKeySize> public_key,
+    base::span<const uint8_t> cc) {
+  std::array<uint8_t, 1 + 32 + 4> data;
+  auto span_writer = base::SpanWriter(base::span(data));
+  span_writer.Skip(1u);
+  span_writer.Write(public_key);
+  span_writer.WriteU32LittleEndian(index);
+
+  DerivedHmacs result;
+
+  data[0] = 0x02;
+  result.z_hmac = crypto::hmac::SignSha512(cc, data);
+
+  data[0] = 0x03;
+  result.cc_hmac = crypto::hmac::SignSha512(cc, data);
+
+  return result;
+}
```

Comment:
can we merge these two functions? And for normal derivation, we can just calculate it from scalar

Reply:
merged